csi/src/findmnt.rs

```diff
@@ -0,0 +1,223 @@
+            if filter == value {
+                results.push(jsonmap_to_hashmap(json_map));
+            }
+        }
+        // If the object has arrays, then the assumption is that they are arrays
+        // of objects.
+        for (_, jsonvalue) in json_map {
+            if jsonvalue.is_array() {
+                filter_findmnt(jsonvalue, filter, results);
+            }
+        }
+    }
+}
+
+/// findmnt executable name.
+const FINDMNT: &str = "findmnt";
+/// findmnt arguments, we only want source, target and filesystem type fields.
+const FINDMNT_ARGS: [&str; 3] = ["-J", "-o", "SOURCE,TARGET,FSTYPE"];
+
+/// Execute the Linux utility findmnt, collect the json output,
+/// invoke the filter function and return the filtered results.
+fn findmnt(
+    params: FindmntFilter,
+) -> Result<Vec<HashMap<String, String>>, DeviceError> {
+    let output = Command::new(FINDMNT).args(&FINDMNT_ARGS).output()?;
+    if output.status.success() {
+        let json_str = String::from_utf8(output.stdout)?;
+        let json: Value = serde_json::from_str(&json_str)?;
+        let mut results: Vec<HashMap<String, String>> = Vec::new();
+        filter_findmnt(&json, &params, &mut results);
+        Ok(results)
+    } else {
+        Err(DeviceError {
+            message: String::from_utf8(output.stderr)?,
+        })
+    }
+}
+
+/// Use the Linux utility findmnt to find the name of the device mounted at a
+/// directory or block special file, if any.
+/// mount_path is the path a device is mounted on.
+pub(crate) fn findmnt_get_devicepath(
+    mount_path: &str,
+) -> Result<Option<String>, DeviceError> {
+    let tgt_filter = FindmntFilter {
+        key: TARGET_KEY,
+        value: mount_path,
+    };
+    match findmnt(tgt_filter) {
+        Ok(sources) => {
+            match sources.len() {
+                0 => Ok(None),
+                1 => {
+                    if let Some(devicepath) = sources[0].get(SOURCE_KEY) {
+                        Ok(Some(devicepath.to_string()))
+                    } else {
+                        Err(DeviceError {
+                            message: "missing source field".to_string(),
+                        })
+                    }
+                }
+                _ => {
+                    // should be impossible ...
+                    warn!(
+                        "multiple sources mounted on target {:?}->{}",
+                        sources, mount_path
+                    );
+                    Err(DeviceError {
+                        message: format!(
+                            "multiple devices mounted at {}",
+                            mount_path
+                        ),
+                    })
+                }
+            }
+        }
+        Err(e) => Err(e),
+    }
+}
+
+/// Use the Linux utility findmnt to find the mount paths for a block device,
+/// if any.
+/// device_path is the path to the device for example "/dev/sda1"
+pub(crate) fn findmnt_get_mountpaths(
```

Comment:
Since we have namespaces, why do we prefix everything with findmnt? why not get_mounts() as it would then be

findmnt::get_mounts() rather then findmnt::findmnt_get_mount_paths() ?

Reply:
Done.